### ISSUE TITLE
inline decorators

### ISSUE THREAD

**Issue by zh99998** [prettier/prettier#2613]
continue from #325, I think it need more discussion

It's really ugly to put them in one line in field definition like in TypeORM.

maybe only decorators without () should be inline?

before

```typescript
@Entity()
export class Board {
  
    @PrimaryGeneratedColumn()
    id: number;
    
    @Column()
    slug: string;
    
    @Column()
    name: string;
    
    @Column()
    theme: string;
    
    @Column()
    description: string;
    
    @OneToMany(type => Topic, topic => topic.board)
    topics: Topic[]
  
}
```

after

```typescript
@Entity()
export class Board {
  @PrimaryGeneratedColumn() id: number;

  @Column() slug: string;

  @Column() name: string;

  @Column() theme: string;

  @Column() description: string;

  @OneToMany(type => Topic, topic => topic.board)
  topics: Topic[];
}
```

**Comment by azz** [comment]
This was fixed way back in #1910

**Prettier 1.7.4**
[Playground link](https://prettier.io/playground/#%7B%22content%22%3A%22%40Entity()%5Cnexport%20class%20Board%20%7B%5Cn%20%20%5Cn%20%20%20%20%40PrimaryGeneratedColumn()%5Cn%20%20%20%20id%3A%20number%3B%5Cn%20%20%20%20%5Cn%20%20%20%20%40Column()%5Cn%20%20%20%20slug%3A%20string%3B%5Cn%20%20%20%20%5Cn%20%20%20%20%40Column()%5Cn%20%20%20%20name%3A%20string%3B%5Cn%20%20%20%20%5Cn%20%20%20%20%40Column()%5Cn%20%20%20%20theme%3A%20string%3B%5Cn%20%20%20%20%5Cn%20%20%20%20%40Column()%5Cn%20%20%20%20description%3A%20string%3B%5Cn%20%20%20%20%5Cn%20%20%20%20%40OneToMany(type%20%3D%3E%20Topic%2C%20topic%20%3D%3E%20topic.board)%5Cn%20%20%20%20topics%3A%20Topic%5B%5D%5Cn%20%20%5Cn%7D%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Afalse%2C%22output2%22%3Afalse%2C%22parser%22%3A%22typescript%22%2C%22printWidth%22%3A80%2C%22semi%22%3Atrue%2C%22singleQuote%22%3Afalse%2C%22tabWidth%22%3A2%2C%22trailingComma%22%3A%22none%22%2C%22useTabs%22%3Afalse%7D%7D)
```sh
--parser typescript
```

**Input:**
```tsx
@Entity()
export class Board {
  
    @PrimaryGeneratedColumn()
    id: number;
    
    @Column()
    slug: string;
    
    @Column()
    name: string;
    
    @Column()
    theme: string;
    
    @Column()
    description: string;
    
    @OneToMany(type => Topic, topic => topic.board)
    topics: Topic[]
  
}
```

**Output:**
```tsx
@Entity()
export class Board {
  @PrimaryGeneratedColumn() id: number;

  @Column() slug: string;

  @Column() name: string;

  @Column() theme: string;

  @Column() description: string;

  @OneToMany(type => Topic, topic => topic.board)
  topics: Topic[];
}

```

**Comment by egeozcan** [comment]
@azz the output is very undesirable for some of us. very hard to see the actual properties when decorators are inlined...

**Comment by emzero** [comment]
Agreed. Who likes inlined decorators? =/

**Comment by pleerock** [comment]
@azz can this please be re-opened? Inline decorators are ugly. I can gather lot of people who will tell the same. Also as you can see lot of people in merged PR were complaining about inline decorators.

**Comment by Hoten** [comment]
Can this styling choice be reevaluated?

**Comment by duailibe** [comment]
Can one of you open a new issue with **real world code examples** that show how this style looks worse.

**Comment by emzero** [comment]
A TypeORM entity with inline decorators as prettier format it
![image](https://user-images.githubusercontent.com/412489/40514798-edd67006-5f80-11e8-9d95-6876db053c08.png)

The same TypeORM entity with decorators in its own line
![image](https://user-images.githubusercontent.com/412489/40514914-40c54008-5f81-11e8-9a06-fff069933af8.png)

If the decorator has parameters then it is in its own line. But if doesn't have parameters it is inlined next to the property. This makes it really hard to spot which are the properties at first sight. And honestly, if you tell me you prefer the first one but there are a lot of us who prefer the second one, then that should be enough reason to make it configurable IMHO.

**Comment by Dragory** [comment]
Just stumbled upon this same issue working with TypeORM entities. Please reconsider this, or add a configurable option - the inlined example above is very hard to parse at a glance.

**Comment by zh99998** [comment]
should we re-open this issue  ?

@azz

**Comment by azz** [comment]
Decorators still aren't part of the language, so I'm not in any hurry to create churn when the spec may change (it actually already has). If we do change it we'll likely get a flood of feedback from people who like it the way it is.

The current behaviour was requested in #1898.

**Comment by egeozcan** [comment]
@azz how is the decision process? what about making it configurable as suggested in the discussion you mentioned?

**Comment by azz** [comment]
Have a read of our [option philosophy](https://prettier.io/docs/en/option-philosophy.html).

**Comment by pleerock** [comment]
> Decorators still aren't part of the language, so I'm not in any hurry to create churn when the spec may change (it actually already has). If we do change it we'll likely get a flood of feedback from people who like it the way it is.

First, even decorators are not "part of the language", they are part of the TypeScript which is completely supported by prettier. So, we can say that it is part of the language.

Second, if even decorators will change they semantically will stay same, otherwise they won't be called decorators anymore.

Third and the most important, if you are not sure how to format this functionality then its better not format it at all. 

What I want to say is that if you are not sure how to properly format this code because its not part of the language then its probably better not to format it at all and let people decide how to format it. I see that some guy decided to format it the way he liked, sent PR and you accepted it, but... its not how it should be done. Prettier is very popular nowadays and this popularity brings big responsibility to it, and each such PR should be taken carefully with enough level of insurance that this change bring a better code and most people going to admit and accept it.

I know, its hard to achieve, because people starting being active only when they get issues, and maybe this is exact case.

I haven't spread a word about this issue in the community, but I already see that my comment above received 20 votes which may be enough to make you think that something is being formatted not the way people actually prefer. Please re-consider PR you accepted that already has more downvotes then upvotes.

**Comment by azz** [comment]
> even decorators are not "part of the language", they are part of the TypeScript 

To the same extent as they are a part of JavaScript. Decorators re behind the `experimentalDecorators` flag in TypeScript. 

> if you are not sure how to format this functionality then its better not format it at all.

This would undermine the value prospect of Prettier. We've had requests to ignore the input code 100% (#2068). We have the `// prettier-ignore` comment as an escape hatch which will allows you to opt-out of formatting a block of code, but we try to ignore the source text as much as possible to reduce ambiguity and confusion. 

> some guy decided to format it the way he liked, sent PR and you accepted it

That "some guy" is @vjeux (#1910) and he is the number one contributor to Prettier. 😄 

> my comment above received 20 votes

Just to put it in perspective, we've got issues that we haven't acted on with 100+ votes (e.g. #3847). We walk a very fine line when making changes as it effects a very large amount of people, and generally the people who like the current behaviour don't go and search for issues to change it. 

I want to clarify my previous comment and say that I actually don't have an opinion on this issue as I've avoided decorators for the most part. _If_ the community at-large was in favour of the change then it would make sense to proceed. However, we need to gather more data and more feedback before making such a change.

**Comment by pleerock** [comment]
> However, we need to gather more data and more feedback before making such a change.

why you did not gather data before implementing current behaviour? All I see is in issues just one-two opinions and issues are even marked as "subjective". I think if you are not sure about some formatting, then its really better not to do it.

> Just to put it in perspective, we've got issues that we haven't acted on with 100+ votes

now count. You have 100 votes for a popular use case. And you have a 25 votes for non-popular, experimental syntax, used only by some people. Don't you think think that value here is already higher then in that 100+ votes issue?

> We walk a very fine line when making changes as it effects a very large amount of people,

exactly! That's why we are worrying here. Current formatting effects us and we don't understand why current changes ware made without walking a very fine line.

**Comment by azz** [comment]
> That's why we are worrying here. Current formatting effects us and we don't understand why current changes ware made without walking a very fine line.

The change you're referring to happened a year ago. `1.4.0` had TS support, `1.4.1` had this change, both were released on the same day. So it wasn't really a _change_ in that sense, more it was early tuning.

**Comment by emzero** [comment]
We use TypeScript in most of our projects and we indeed use decorators a lot. We have now decided to ditch Prettier and go back to our previous formatter because of this. 

If in the future this gets addressed by either making it configurable or by don't formatting decorators at all, someone please let me know. Thanks.

**Comment by babeal** [comment]
This is another big one for us.  Inlined decorators are absolutely horrible IMO.  Source code reading has been hurt by 78.567% by this ;-)

**Comment by zh99998** [comment]
so pick one of them would solve problem

1. all decorators should not inline
2. decorators with bracket should not inline

**Comment by suchipi** [comment]
To make TypeORM better, let's detect classes decorated with `@Entity()` and not inline any of their decorators.

**Comment by paulparton** [comment]
TypeORM is just one example of this. Our case is in Angular with component input/output decorators. If there is going to be a change to make this configurable / optional please don't just do it for one specific case.

**Comment by duailibe** [comment]
@paulparton @pleerock Can you share real world examples of code with before/after that were affected by Prettier? Otherwise it’ll be realy hard to come up with ideas

**Comment by pleerock** [comment]
@duailibe its not related to any specific library like TypeORM, so treat examples as generic since they only use decorators, not typeorm or other-library specific functionality/feature/style or anything else. Just decorators. Just take a look on a original post and compare "before" and "after". Most of people admin that before is better. Plus try to add more decorators on properties, try to add more decorators with longer names, no matter if its from some library or you own created decorators - the result is ugly and less readable if they were each in a separate row.

**Comment by suchipi** [comment]
@pleerock @paulparton I have no personal issue changing it but I want to fully understand the issue before making a change, since prettier is used by so many people, and it's hard to do that without code examples.

**Comment by suchipi** [comment]
Looks like #325 also mentions mobx:

**Prettier 1.13.4**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAtgBwgJxgAmHwgCMBnOHANwEMSAbOfAX3wDMcJ18AdEdUgA8+Abh5RxYejTJl8AeRwATSgBlUUJsHH58AAVIVqdRjvyYcqMHCT4oAV3QlK+ALz4ADGIlRdB8pS0DHBmNAL2sLYOTi7uAIzeZpBQZDA49mAwuAAUFlZwAJSEZrowABaoZAB0edZu5pbW3rrM4mZ6kFj28EpmAOZwBFkwNPTZRUQl+DiD9ji+5ZU1jUwAVPiL1WEQETDNLOKtUCAANCAQmDCo0GTIoDQ4nADuAAoPCLcoNFQQqEqnIBIOBoYAA1oMAMqYEEaPrINL2OBnDRGGAvYF9dA0ZBsUYUM4AKzIggAQsCwZCwnB1JocXikSBoTgjMhAXQAJ70aAAvKwADqf3KyAAHB4zhYIBQ+cDMKyLHAjFQ4ACZgBHeyoGbomiY7FIXH0fEgCjoVDw9IMsiwxgARXsEHgdMNDJGJAFSiFSAATGc0jRUPRYQBhLhY1lQaDKs72CgAFTonwNFGYzCAA)
```sh
--parser babylon
```

**Input:**
```jsx
import { observable } from "mobx";

class OrderLine {
  @observable
  price: number = 0;

  @observable
  amount: number = 1;

  constructor(price) {
    this.price = price;
  }

  @computed
  get total() { 
    return this.price * this.amount;
  }
}

```

**Output:**
```jsx
import { observable } from "mobx";

class OrderLine {
  @observable price: number = 0;

  @observable amount: number = 1;

  constructor(price) {
    this.price = price;
  }

  @computed
  get total() {
    return this.price * this.amount;
  }
}

```

**Comment by paulparton** [comment]
@duailibe @suchipi Heya! sure thing, here is some Angular code.

Input

<img width="714" alt="screen shot 2018-06-29 at 8 48 18 am" src="https://user-images.githubusercontent.com/5544897/42119762-006f68f0-7c53-11e8-81b0-a4402ce1e2aa.png">

Output

<img width="698" alt="screen shot 2018-06-29 at 8 53 07 am" src="https://user-images.githubusercontent.com/5544897/42119768-0b368cf0-7c53-11e8-9801-772974ec3690.png">

**Comment by drudv** [comment]
> why you did not gather data before implementing current behaviour? All I see is in issues just one-two opinions and issues are even marked as "subjective". I think if you are not sure about some formatting, then its really better not to do it.

I think this should be a part of pettier philosophy: implement a formatter only after gathering data.

**Comment by egeozcan** [comment]
Established similar structures from other programming languages:

c# Attributes

![attributescsharp](https://user-images.githubusercontent.com/185294/42186882-7904c434-7e4e-11e8-83ee-c12322dfb182.jpg)

java annotations:

![javaannotations](https://user-images.githubusercontent.com/185294/42186959-be8b104e-7e4e-11e8-96b4-da96580dbdae.png)
[Google style guide for them](https://google.github.io/styleguide/javaguide.html#s4.8.5-annotations)

**Comment by suchipi** [comment]
I'm convinced- let's change it

**Comment by egeozcan** [comment]
There's also the decorators proposal itself: https://github.com/tc39/proposal-decorators

There, they are also on their own lines.

**Comment by j-f1** [comment]
> Is this officially planned to be reverted? If yes, when?

We don’t really have an official roadmap. It’ll happen when someone (a maintainer or an external contributor) opens a PR :)

**Comment by scott-ho** [comment]
@j-f1 @suchipi plz consider about this issue https://github.com/prettier/prettier/issues/4924
